### PR TITLE
AdvectionBC and test

### PR DESCRIPTION
**Pull request by snschune**

### CODE REVIEW

**Comment by snschune** [comment]
closes #13283

File: modules/navier_stokes/src/bcs/AdvectionBC.C [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                               "The components of the velocity vector up to problem dimension");
+  return params;
+}
+
+AdvectionBC::AdvectionBC(const InputParameters & parameters)
+  : IntegratedBC(parameters), _dim(_mesh.dimension()), _outflow(getParam<bool>("outflow"))
+{
+  // check if # components matches mesh's dim
+  if (_dim != coupledComponents("velocity_vector"))
+    mooseError("Number of components of velocity_vector must match mesh dimension");
```
Comment (permcody) [inline modules/navier_stokes/src/bcs/AdvectionBC.C]: paramError()

File: modules/navier_stokes/src/bcs/AdvectionBC.C [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    _velocity[j] = &coupledValue("velocity_vector", j);
+}
+
+Real
+AdvectionBC::computeQpResidual()
+{
+  RealVectorValue vel;
+  for (unsigned int j = 0; j < _dim; ++j)
+    vel(j) = (*_velocity[j])[_qp];
+  if (!((vel * _normals[_qp] > 0) ^ _outflow))
```
Comment (permcody) [inline modules/navier_stokes/src/bcs/AdvectionBC.C]: Bitwise XOR? You probably want a comment here!

File: modules/navier_stokes/src/bcs/AdvectionBC.C [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+template <>
+InputParameters
+validParams<AdvectionBC>()
+{
+  InputParameters params = validParams<IntegratedBC>();
+  params.addClassDescription("Boundary conditions for outflow/outflow of advected quantities:"
+                             "\n phi * velocity * normal, where phi is the advected quantitiy");
+  params.addParam<bool>(
+      "outflow", true, "Determines if this BC is applied on outflow or outflow BCs");
```
Comment (permcody) [inline modules/navier_stokes/src/bcs/AdvectionBC.C]: Typo

**Review by permcody — CHANGES_REQUESTED** [review]

**Comment by snschune** [comment]
Should this BC even be allowed on inflow BCs? ping @YaqiWang

File: modules/navier_stokes/src/bcs/AdvectionBC.C (hunk 2 above)
Comment (permcody) [reply modules/navier_stokes/src/bcs/AdvectionBC.C]: This should be equivalent and more clear:
`if ((vel * _normals[_qp] > 0) == _outflow)`

**Review by permcody — APPROVED** [review]

**Pull request merged**